horovod/run/common/util/safe_shell_exec.py

```diff
@@ -69,75 +74,104 @@ def prepend_context(line, rank, prefix):
-    # Make a pipe for the subprocess stdout/stderr.
-    (stdout_r, stdout_w) = os.pipe()
-    (stderr_r, stderr_w) = os.pipe()
 
-    # Make a pipe for notifying the child that parent has died.
-    (r, w) = os.pipe()
+def _exec_middleman(command, env, exit_event, stdout, stderr, rw):
+    stdout_r, stdout_w = stdout
+    stderr_r, stderr_w = stderr
+    r, w = rw
+
+    # Close unused file descriptors to enforce PIPE behavior.
+    stdout_r.close()
+    stderr_r.close()
+    w.close()
+    os.setsid()
 
-    middleman_pid = os.fork()
-    if middleman_pid == 0:
-        # Close unused file descriptors to enforce PIPE behavior.
-        os.close(w)
-        os.setsid()
+    executor_shell = subprocess.Popen(command, shell=True, env=env,
+                                      stdout=stdout_w, stderr=stderr_w)
 
-        executor_shell = subprocess.Popen(command, shell=True, env=env,
-                                          stdout=stdout_w, stderr=stderr_w)
+    def kill_executor_children_on_event():
+        exit_event.wait()
+        terminate_executor_shell_and_children(executor_shell.pid)
 
-        sigterm_received = threading.Event()
+    in_thread(kill_executor_children_on_event)
 
-        def set_sigterm_received(signum, frame):
-            sigterm_received.set()
+    def kill_executor_children_if_parent_dies():
+        # This read blocks until the pipe is closed on the other side
+        # due to parent process termination (for any reason, including -9).
+        os.read(r.fileno(), 1)
+        terminate_executor_shell_and_children(executor_shell.pid)
 
-        signal.signal(signal.SIGINT, set_sigterm_received)
-        signal.signal(signal.SIGTERM, set_sigterm_received)
+    in_thread(kill_executor_children_if_parent_dies)
 
-        def kill_executor_children_if_parent_dies():
-            # This read blocks until the pipe is closed on the other side
-            # due to the process termination.
-            os.read(r, 1)
-            terminate_executor_shell_and_children(executor_shell.pid)
+    exit_code = executor_shell.wait()
+    if exit_code < 0:
+        # See: https://www.gnu.org/software/bash/manual/html_node/Exit-Status.html
+        exit_code = 128 + abs(exit_code)
 
-        in_thread(kill_executor_children_if_parent_dies)
+    sys.exit(exit_code)
 
-        on_event(sigterm_received, terminate_executor_shell_and_children, args=(executor_shell.pid,))
 
-        exit_code = executor_shell.wait()
-        os._exit(exit_code)
+def _create_event(ctx):
+    # We need to expose this method for internal testing purposes, so we can mock it out to avoid
+    # leaking semaphores.
+    return ctx.Event()
+
+
+def execute(command, env=None, stdout=None, stderr=None, index=None, events=None):
+    ctx = multiprocessing.get_context('spawn') if _PY3 else multiprocessing
+
+    # When this event is set, signals to the middleman that it should terminate its children
+    # and exit.
+    exit_event = _create_event(ctx)
```

Comment:
we can't use `threading.Event` here because `middleman` is a separate thread (IPC), right?

Reply:
We need to use a `multiprocessing.Event` because the middleman is a separate process.  The multiprocessing event is identical to the threading event, except that works over separate processes instead of threads.  However, kill -9 on a subprocess will do really bad things with global state when using an Event like this, which is why we stub it out into a mockable function.